### PR TITLE
Support SQLite Database

### PR DESCRIPTION
**Pull request by vzool**
#27

### CODE REVIEW

**Comment by zigazajc007** [comment]
Thanks for your pull request, I will review it in few days.

File: api/php.ini [hunk 1]
```diff
@@ -144,7 +144,7 @@ session.referer_check =
 session.cache_limiter = nocache
 session.cache_expire = 180
 session.use_trans_sid = 0
-session.sid_length = 48
+session.sid_length = 64
```
Comment (vzool) [inline api/php.ini:147]: Why not max it out to the maximum possible value (256) to increase the security level?
"[Longer session ID is harder to guess](https://www.php.net/manual/en/session.configuration.php#ini.session.sid-length)".  🤗

**Comment by zigazajc007** [comment]
Should we use SQLite as default database provider for easier installation?

So we can also remove database configuration from installers, but still leave it in .env.example for tinkerers.

**Comment by vzool** [comment]
> Should we use SQLite as default database provider for easier installation?

Yes, I think so. Because by default the application with SQLite will be standalone without any external requirements.
This will ease the automated test later, but the tests will cover all types of supported databases.
But, when the configuration changes it will be overwritten directly.
This [PR](https://github.com/Rabbit-Company/Passky-Server/pull/28) contains an "[Auto-Migration](https://github.com/vzool/Passky-Server/blob/sqlite-support/api/src/Settings.php#L131-L222)" which it will check for the tables if not exists it will be created automatically.
At the database level for SQLite it is just a file which will be created by [PDO Engine](https://www.php.net/manual/en/ref.pdo-sqlite.php), but in MySQL it will create the database itself if it doesn't exist, then create all the tables.

> So we can also remove database configuration from installers, but still leave it in .env.example for tinkerers.

I don't recommend removing anything from the configuration that is generated by installers, I like to help tinkerers, so even "[installer.sh](https://github.com/vzool/Passky-Server/blob/sqlite-support/installer.sh)" and "[installerGUI.sh](https://github.com/vzool/Passky-Server/blob/sqlite-support/installerGUI.sh)" with leave all settings for tinkerers. And `DATABASE_ENGINE` will make the auto switch between them.

**Comment by vzool** [comment]
You know after the deployment is done and if the database name changed for all `DATABASE_ENGINE`, then it will create the database and the tables automatically. 🤗

**Comment by zigazajc007** [comment]
Will also include the documentation in this pull request if you don't mind.

**Comment by vzool** [comment]
> Will also include the documentation in this pull request if you don't mind.

New documentations structure looks very nice, that's will open the doors for a whole new expand ideas. 😍✌️

**Comment by zigazajc007** [comment]
> > Will also include the documentation in this pull request if you don't mind.
> 
> The new documentation structure looks very nice, which will open the doors for whole new expand ideas. heart_eyesv

Will also remove almost everything from the main README and make it more organized. There is currently too much useless text.

Your PR looks great, but I still need to think about Auto-Migration.

**Comment by zigazajc007** [comment]
Database structure also won't change frequently and if it will, then only on big updates.

Current docker-compose file will fetch images from latest tag. This should be change to a specific version or to a current tag 8 (Will change the workflow to also deploy images under tag with first version number). So `docker-compose pull` will only pull images that are under tag 8 (no breaking changes included).

And for a breaking change update user will need to manually switch version and there we can put a comment / warning to read the upgrade guide.

Currently if user use existing .env file and don't go thru installer there will be an error:
```yaml
Step 5/19 : COPY .tmp/database.mysql.sql /var/www/html/
COPY failed: file not found in build context or excluded by .dockerignore: stat .tmp/database.mysql.sql: file does not exist
ERROR: Service 'passky-api' failed to build : Build failed
```
the same error will happen if user wouldn't use the installer and change settings from .env.example manually.

So personally I wouldn't include any other action in installers than just writing to .env file. The less we mess with the file system the better it will be. Also when creating docker images it will be required to have database inside .tmp folder. Would rather see to include `database.sqlite.sql` file in `api/` folder (It's not needed in database folder) or to hard-code the table creation inside Settings.php or new .php file.

This would also allow us to make a simple website where user can answer questions and it will produce him with text that he can copy paste in `api/` folder, this will also be a first step to allow us to avoid using any terminal for shared hosting.

Would also like to hear your thoughts on that.

**Comment by vzool** [comment]
> Will also remove almost everything from the main README and make it more organized. There is currently too much useless text.

Wonderful. 

> Your PR looks great, but I still need to think about Auto-Migration.

Yes, of course. We need to rethink about Auto-Migration, especially after moving database schema inside `api/`

**Comment by vzool** [comment]
> Database structure also won't change frequently and if it will, then only on big updates.

Agree

> Current docker-compose file will fetch images from latest tag. This should be change to a specific version or to a current tag 8 (Will change the workflow to also deploy images under tag with first version number). So `docker-compose pull` will only pull images that are under tag 8 (no breaking changes included).
> 
> And for a breaking change update user will need to manually switch version and there we can put a comment / warning to read the upgrade guide.
> 
> Currently if user use existing .env file and don't go thru installer there will be an error:
> 
> ```yaml
> Step 5/19 : COPY .tmp/database.mysql.sql /var/www/html/
> COPY failed: file not found in build context or excluded by .dockerignore: stat .tmp/database.mysql.sql: file does not exist
> ERROR: Service 'passky-api' failed to build : Build failed
> ```
> 
> the same error will happen if user wouldn't use the installer and change settings from .env.example manually.
> 
> So personally I wouldn't include any other action in installers than just writing to .env file. The less we mess with the file system the better it will be. Also when creating docker images it will be required to have database inside .tmp folder. Would rather see to include `database.sqlite.sql` file in `api/` folder (It's not needed in database folder) or to hard-code the table creation inside Settings.php or new .php file.

Totally agree, let's respect [Single-responsibility Principle (SRP)](https://en.wikipedia.org/wiki/Single-responsibility_principle) for `.env` generators, and make all database schema scripts inside `api/`.

> This would also allow us to make a simple website where user can answer questions and it will produce him with text that he can copy paste in `api/` folder, this will also be a first step to allow us to avoid using any terminal for shared hosting.
>
> Would also like to hear your thoughts on that.

I will push new updates.

**Comment by zigazajc007** [comment]
`database/` folder is the root folder for passky-database docker image creation. So in this folder we don't need sqlite file.

If anyone would want to use their own database it will most likelly check folder called database and see .sql file. So he would just need to copy paste the table creation queries.

I would put sqlite file only in api/ folder, so in case database isn't configured it will just use SQLite.

**Comment by vzool** [comment]
> `database/` folder is the root folder for passky-database docker image creation. So in this folder we don't need sqlite file.
> 
> If anyone would want to use their own database it will most likelly check folder called database and see .sql file. So he would just need to copy paste the table creation queries. (Would also write the documentation)
> 
> I would put sqlite file only in `api/` folder, so in case database isn't configured it will just use SQLite.

I'm not sure, who will use the database without the `api` itself!
So, I think `api` should be responsible for all database schema scripts. 🤗

**Comment by zigazajc007** [comment]
Do you think it's good to use CREATE DATABASE statement?

For example Wordpress or most other frameworks will require you to create database on your own and create the user that only has permissions inside this database. We probably shouldn't use CREATE DATABASE statement as for this to work user would need to have all permissions which usually isn't good.

It's great for auto creating tables if they do not exists, but I'm not sure for the database.

If you use external database providers like [PlanetScale](https://planetscale.com) (We are using it for public Passky instances) you would need to create the database first in order to get user and password to connect as this user will only work on this specific database.

**Comment by vzool** [comment]
Good point, the `CREATE DATABASE` statement is in the exception of [try/catch block](https://github.com/vzool/Passky-Server/blob/sqlite-support/api/src/Settings.php#L134-L162), when the database is unknown the `CREATE DATABASE` statement will run. Otherwise, any predefined database will not raise any exception so it will be completely ignored.

**Comment by vzool** [comment]
Even there is a [condition](https://github.com/vzool/Passky-Server/blob/sqlite-support/api/src/Settings.php#L142-L144) to track a specific exception which is "**unknown database**" only. 🤗

**Comment by zigazajc007** [comment]
For SQLite this migration system should be fine. But for mysql (production) system we are paying for queries / row reads. Imagine if every time user calls the API endpoint it will first need to check for migration process and then execute their requested query. So it would increase latency by a lot if database isn't located on the same server.

So if 100 people are using Passky Server at the same time, then 100 additional queries will be executed just to check for  migration. While not having tables created is only one time issue, so checks doesn't need to be performed every request.

For MySQL I would still prefer to not have a migration system. SQLite should be perfect for self-hosters, but for production use you will most likely use MySQL or MariaDB and if you choose to use MySQL or MariaDB you will probably also know how to create tables.

**Comment by vzool** [comment]
Great vision is clear now. 🤗
We should be minimalist for any working environment.
So, I think the best plan for any database engine is "**Migration system must work only once in a lifetime**" 😋✌️

**Comment by vzool** [comment]
In order to implement "**once in a lifetime**" algorithm for the migration system, I used a [File-Based Sentinel Value](https://en.wikipedia.org/wiki/Sentinel_value). 🤗

**Comment by vzool** [comment]
@zigazajc007 Is this [PR](https://github.com/Rabbit-Company/Passky-Server/pull/28) still under review?

**Comment by zigazajc007** [comment]
I have started working on a project called Bloggy around 3 weeks ago and was programming 10 to 15 hours per day every day. The project is almost finished, so I took Friday and Saturday off. I will most likelly took Sunday off as well and start working again on Monday.

For Bloggy I just wanted to use Plausible for analytics and use their API to generate statistics for creators, but recently I have decided to make my own analytics, so I can only include features that I need. I have managed to make it 5x less bloated than Plausible and JS contains only 33 lines of code in comparison to 161 from Plausible.

I have already programmed more in 2023 than in whole 2021 and will soon surpass year 2022 in about a month or two. So I'm looking forward for the year 2023.

**Comment by vzool** [comment]
The future is bright ... Take your time my friend, and Good luck with 2023 😋✌️

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,10 @@
...
+api/databases
+api/data.json
 api/api-limiter.json
 api/tokens.json
 api/vendor
 api/composer.lock
-node_modules
\ No newline at end of file
+node_modules
+.migration.php
```
Comment (vzool) [inline .gitignore:10]: Not required anymore.

File: api/src/Settings.php [hunk 3]
```diff
@@ -1,5 +1,18 @@
 <?php
 
+define('ROOT', realpath(__DIR__ . '/..'));
+define('PUBLIC', __DIR__);
+define('DATA_FILE', ROOT . '/data.json');
+define('MIGRATION_FILE', ROOT . '/.migration.php');
```
Comment (vzool) [inline api/src/Settings.php:6]: Not required anymore.

File: server/src/Settings.php [hunk 4]
```diff
@@ -89,7 +89,7 @@ public static function getDBPassword() : string{
 	}
 
 	public static function getDBSSL() : bool{
-		return in_array(getenv('MYSQL_SSL', true), ['true', '1'], true) && @filesize(self::getDBSSLCA());
```
Comment (vzool) [inline server/src/Settings.php]: **Notes**:
- If SSL files are not found then `MYSQL_SSL` should be disabled by default.
- `MYSQL_SSL` when enabled its value in docker is `"true"` but outside is `"1"`.

**Review by vzool — COMMENTED** [review]
That's all.

File: server/src/Settings.php (hunk 4 above)
Comment (zigazajc007) [reply server/src/Settings.php]: Even if I enable SSL with default path to certificates, it doesn't get enabled. So this doesn't work as it should.

Comment (vzool) [reply server/src/Settings.php]: I think that I found an issue with this outside of the docker, so I made the proper changes. ^_^

Comment (zigazajc007) [reply server/src/Settings.php]: I also don't think this should fall to false if file is incorrect or it doesn't exists as then user will think that they are using secure connection, but in reality it has been silently disabled in the back end.

Comment (zigazajc007) [reply server/src/Settings.php]: In this case it would be better for the Error to be thrown or connection to a database unsuccessful.

Comment (vzool) [reply server/src/Settings.php]: OK

**Pull request merged**